File: ApplicationUpdater/updater.py

import requests
import json
import zipfile
import os
import shutil
import time
import sys
import zc.lockfile

# Make sure program is only running once
try:
    lock = zc.lockfile.LockFile('lock')
except:
    print("Can't check for updates: Another process is still running...")
    sys.exit(-1)

automaticUpdaterRelease = "AutomaticUpdaterRelease.zip"

def getLatestGithubRelease(user, repository):
    return requests.get("https://api.github.com/repos/" + user + "/" + repository + "/releases/latest")

def download_file(url):
    local_filename = url.split('/')[-1]
    # NOTE the stream=True parameter below
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        with open(local_filename, 'wb') as f:
            for chunk in r.iter_content(chunk_size=8192): 
                # If you have chunk encoded response uncomment if
                # and set chunk_size parameter to None.
                #if chunk: 
                f.write(chunk)
    return local_filename

def downloadUpdate(j):
    print("Downloading " + automaticUpdaterRelease + "...")
    try:
        assets = j["assets"]
        for asset in assets:
            if (asset["name"] == automaticUpdaterRelease):
                url = asset["browser_download_url"]

                download_file(url)
                print("Successfully downloaded " + automaticUpdaterRelease + "\n")   
                return 1
    except:
        print("Failed to download update\n")
        return 0

def unzipUpdate():
    print("Unzipping update...")
    try:
        with zipfile.ZipFile(automaticUpdaterRelease, 'r') as zip_ref:
            zip_ref.extractall("update")
        
        print("Successfully unzipped update\n")
        return 1
                    
    except:
        print("Failed to unzip update\n")
        return 0

def updateApplication():
    print("Applying update...")
    outdatedSet = 0
    
    try:
        # Make the outdated folder empty
        if (os.path.exists("outdated")):
            print("\tDeleting outdated folder...")
            try:
                # First, delete all .exe files to prevent deleting 
                # resource files while an exe is running
                for file in os.listdir("outdated"):
                    if (file.endswith(".exe")):
                        os.remove(os.path.join("outdated", file))
                # Next, delete the entire folder
                shutil.rmtree("outdated")
                print("\tDone")
            except:
                print("\tDeletion failed")
                raise
        
        # Move currently latest file to outdated
        if (os.path.exists("latest")):
            print("\tMoving latest files to outdated...")
            try:
                os.mkdir("outdated")
                # Now move all files to outdated
                for file in os.listdir("latest"):
                    shutil.move(os.path.join("latest", file), os.path.join("outdated", file))
                print("\tDone migrating to outdated")
            except:
                print("\tFailed migrating old files")
                raise
        else:
            print("\t'latest' folder does not exist, creating...")
            try:
                os.mkdir("latest")
                print("\tDone")
            except:
                print("\tFailed to create latest directory")
                raise

        outdatedSet = 1
        # Now all files are outdated, if an error
        # occurs now, they need to be recovered

        # Now move the new files to latest
        if (os.path.exists("update/app")):
            print("\tCopying new files to latest...")
            try:
                for file in os.listdir("update/app"):
                    shutil.move(os.path.join("update/app", file), os.path.join("latest", file))
                print("\tSuccessfully replaced application files")
            except:
                print("\tFailed copying files")
                raise
        
        # Main update done, update installers now
        print("\tApplication update successful, updating installers now")

        # Check if updates are there
        if (os.path.exists("update/updater.exe") and os.path.exists("update/uninstall.exe")):
            # Now apply them
            print("\tReplacing updater and uninstaller...")
            try:
                if os.path.exists("uninstall.exe"):
                    shutil.move("uninstall.exe", "uninstall_outdated.exe")
                shutil.move("update/uninstall.exe", "uninstall.exe")
                if os.path.exists("updater.exe"):
                    shutil.move("updater.exe", "updater_outdated.exe")
                shutil.move("update/updater.exe", "updater.exe")
                print("\tSuccessfully replaced installers")
            except:
                try:

                    shutil.move("updater_outdated.exe", "updater.exe")
                    shutil.move("uninstall_outdated.exe", "uninstall.exe")
                except:
                    pass
                print("\tFailed to replace installers")
                raise
        else:
            print("\tCan't find new installer files, ignoring them")
        print("Update successful\n")
        return 1
                    
    except:
        print("\nFailed applying the update\n")

        if outdatedSet:
            print("Trying to recover...")
            # Make latest folder empty and copy all outdated files to latest
            try:
                if os.path.exists("latest"):
                    shutil.rmtree("latest")
                    time.sleep(0.1)
                os.mkdir("latest")
                time.sleep(0.1)
                for file in os.listdir("outdated"):
                    shutil.move(os.path.join("outdated", file), os.path.join("latest", file))
                shutil.rmtree("outdated")
                print("Successfully recovered\n")
            except:
                print("Failed to recover\n")

        return 0


def deleteOutdatedFiles():

    # Make the outdated folder empty
    if (os.path.exists("outdated")):
        print("Deleting outdated folder...")
        try:
            # First, delete all .exe files to prevent deleting
            # resource files while an exe is running
            for file in os.listdir("outdated"):
                if (file.endswith(".exe")):
                    os.remove(os.path.join("outdated", file))
            # Next, delete the entire folder
            shutil.rmtree("outdated")
            print("Done")
        except:
            print("Deletion failed")
            raise

    # Delete outdated updater
    if os.path.exists("updater_outdated.exe"):
        print("Deleting outdated updater...")
        try:
            os.remove("updater_outdated.exe")
            print("Done")
        except:
            print("Failed, not significant")

    # Delete outdated uninstaller
    if os.path.exists("uninstall_outdated.exe"):
        print("Deleting outdated uninstaller...")
        try:
            os.remove("uninstall_outdated.exe")
            print("Done")
        except:
            print("Failed, not significant")


def deleteFiles():
    print("Deleting no longer needed files and folders...")

    # Deleting update folder
    if os.path.exists("update"):
        try:
            print("Deleting update temp folder...")
            shutil.rmtree("update")
            print("Done")
        except:
            print("Failed, not significant")

    # Deleting zip file
    if os.path.exists(automaticUpdaterRelease):
        print("Deleting old zip file...")
        try:
            os.remove(automaticUpdaterRelease)
            print("Done")
        except:
            print("Failed, not significant")





deleteOutdatedFiles()

# Find installed version
versionInstalled = ""
try:
    with open("version", "r") as versionFile:
        versionInstalled = versionFile.readlines()[0]
except:
    print("Failed to read installed version")


try:
    release = getLatestGithubRelease("HerrNamenlos123", "TechnicalSketcher")
    if (release.ok):

        j = json.loads(release.content)
        latestVersion = j["tag_name"]

        if (latestVersion != versionInstalled): # Latest version is different from the installed one, update now
            try:
                if (versionInstalled != ""):
                    print("Updating from " + versionInstalled + " to " + latestVersion)
                else:
                    print("Updating to " + latestVersion)
                print("")

                if not downloadUpdate(j):
                    raise ""
                
                if not unzipUpdate():
                    raise ""
                
                if not updateApplication():
                    raise ""
                
                deleteFiles()
                if (versionInstalled != ""):
                    print("\nSuccessfully updated from " + versionInstalled + " to " + latestVersion)
                else:
                    print("\nSuccessfully updated to " + latestVersion)
                print("")
                try:
                    file = open("version","w")
                    file.write(latestVersion) 
                    file.close() 
                except:
                    print("Failed to save current version, application may be updated again")
            except:
                print("Update failed, cleaning files...")
                deleteFiles()
        else:
            print("Up to date: Version " + versionInstalled)

    else:
        print("Failed to access Github release")
except:
    print("Failed to parse Github response")

File: src/Updater.cpp

#include "pch.h"
#include "Updater.h"
#include "Battery/AllegroDeps.h"

/*
 * In November 2021, the installer system was changed from Visual Studio installer projects 
 * to the WiX Toolchain. 
 *  
 * In older versions, there were updater.exe and uninstall.exe. Back then the updates were done
 * by update.exe, which was executed by the TechnicalSketcher. Now this is not needed anymore as
 * all updates are done by the TechnicalSketcher itself.
 * 
 * On existing installations, updater.exe is simply ignored now, it will not do anything as long 
 * as it is not executed and uninstall.exe is also not touched by newer versions, so it can be 
 * uninstalled properly with the original installer.
*/


#define UPDATE_MAX_ATTEMPTS 10		// Attempts for downloading
#define UPDATE_SLEEP_TIME 10		// Seconds to sleep between download attempts

#define TEMP_DIRECTORY "update/"
#define OUTDATED_DIRECTORY "outdated/"


std::shared_ptr<spdlog::logger> logger;

void DeleteOutdatedFiles(const std::string& rootPath) {
	if (Battery::DirectoryExists(rootPath + OUTDATED_DIRECTORY) && Battery::RemoveDirectory(rootPath + OUTDATED_DIRECTORY)) {
		logger->info("Outdated files were deleted");
	}
}

std::pair<std::string, std::string> GetLatestGithubRelease(const std::string& url, int attempt = 0) {

	logger->info("Accessing api.github.com...");

	auto res = Battery::GetHttpRequest(url);
	if (!res.has_value()) {
		logger->info("Connection failed");

		if (attempt < UPDATE_MAX_ATTEMPTS) {
			Battery::Sleep(10);
			logger->info("Trying again");
			return GetLatestGithubRelease(url, attempt + 1);
		}
		else {
			logger->info("Too many attempts, aborting update process");
			return std::make_pair("", "");
		}
	}

	// Connection was valid
	std::string latestVersion;
	std::string downloadLink;
	try {
		nlohmann::json api = nlohmann::json::parse(res->body);
		latestVersion = api["tag_name"];

		for (nlohmann::json asset : api["assets"]) {
			if (asset["name"] == UPDATE_FILENAME) {
				downloadLink = asset["browser_download_url"];
			}
		}
	}
	catch (...) {}

	if (latestVersion.empty() || downloadLink.empty()) {
		logger->info("Failed to parse the JSON response");
		return std::make_pair("", "");
	}

	logger->info("Github response is valid");
	return std::make_pair(latestVersion, downloadLink);
}

bool DownloadUpdate(const std::string& url, const std::string& targetFile, int attempt = 0) {
	size_t cnt = 0;
	logger->info("Downloading {} to {}", url, targetFile);
	if (!Battery::DownloadUrlToFile(url, targetFile, true,
		[&](uint64_t progress, uint64_t total) {				// This is a lambda function!
			Navigator::GetInstance()->updateProgress = (double)progress / total;
			return true;
		}
	)) {
		logger->info("Download failed");

		if (attempt < UPDATE_MAX_ATTEMPTS) {
			Battery::Sleep(UPDATE_SLEEP_TIME);
			logger->info("Trying again...");
			DownloadUpdate(url, targetFile, attempt + 1);
		}
		else {
			logger->info("Too many attempts, stopping download");
			return false;
		}
	}

	return true;
}

void CleanUpdateFiles(const std::string& rootPath) {
	Battery::RemoveDirectory(rootPath + TEMP_DIRECTORY);
	Battery::RemoveFile(rootPath + UPDATE_FILENAME);
}

void RunUpdater() {

	Battery::Sleep(2);

	try {

		// This is the root directory in the installed localappdata folder
		//
		// - %localappdata%
		//   - TechnicalSketcher/  <-- rootdir
		//     - latest/
		//       - TechnicalSketcher.exe
		//       - icon.ico
		//	   - (outdated/)  - These two are used for updating
		//     - (update/)   
		//     - (updater.exe)	  - These two are there in existing installations of an older version
		//     - (uninstall.exe)  - (Not needed anymore)
		//
		std::string installPath = Battery::GetExecutableDirectory();
		std::string rootPath = Battery::GetParentDirectory(installPath);
		std::string installDir = Battery::GetFilename(installPath);
		Navigator::GetInstance()->restartExecutablePath = Battery::GetExecutablePath();

		// Don't update when the parent folder is called 'outdated': It's an old executable
		if (installDir == "outdated") {
			return;
		}

		// Create the lockfile, only one process can update the application at a time
		std::unique_ptr<Battery::Lockfile> lockfile;
		try {
			lockfile = std::make_unique<Battery::Lockfile>(rootPath + "lock");
		}
		catch (...) {
			return;
		}

		// Setup logging
		std::string logfile = rootPath + UPDATER_LOG_FILENAME;
		Battery::RemoveFile(logfile);
		logger = spdlog::basic_logger_mt(UPDATER_LOGGER_NAME, logfile, true);
		logger->set_level(BATTERY_LOG_LEVEL_TRACE);




		// Delete old versions
		DeleteOutdatedFiles(rootPath);



		// Check if the application can be updated

		// First, get the currently installed version
		std::string version = Navigator::GetInstance()->GetApplicationVersion();
		logger->info("Installed version is {}", version);

		// Now check the latest version on Github
		std::pair<std::string, std::string> release = GetLatestGithubRelease(UPDATE_PACKAGE_URL);
		std::string latestVersion = release.first;
		std::string downloadLink = release.second;

		if (!latestVersion.empty()) {
			logger->info("Latest version available on GitHub is {}", latestVersion);
		}
		else {
			logger->info("Could not retrieve latest version tag from GitHub, aborting update process...");
			return;
		}



		// Check if update is necessary
		if (version == latestVersion) {
			logger->info("TechnicalSketcher is up-to-date, no update necessary");
			return;
		}
		else {
			logger->info("TechnicalSketcher can be updated, starting update process...");
		}

		// Now the updating process is official, notify the user
		Navigator::GetInstance()->updateStatus = UpdateStatus::INITIALIZING;
		Navigator::GetInstance()->updateProgress = 0.0;
		Battery::Sleep(0.5);

		// Download the update
		Navigator::GetInstance()->updateStatus = UpdateStatus::DOWNLOADING;
		logger->info("Downloading update package...");
		if (!DownloadUpdate(downloadLink, rootPath + UPDATE_FILENAME)) {
			logger->info("Failed to download update package, aborting update process...");
			Navigator::GetInstance()->timeSincePopup = Battery::GetRuntime();
			Navigator::GetInstance()->updateStatus = UpdateStatus::FAILED;
			return;
		}
		logger->info("Download was successful");

		// Unzip the update
		Navigator::GetInstance()->updateStatus = UpdateStatus::EXTRACTING;
		logger->info("Unzipping the update package...");
		if (!Battery::ExtractArchive(rootPath + UPDATE_FILENAME, rootPath + TEMP_DIRECTORY)) {
			logger->info("Failed to unzip update package, aborting update process...");
			Navigator::GetInstance()->timeSincePopup = Battery::GetRuntime();
			Navigator::GetInstance()->updateStatus = UpdateStatus::FAILED;
			CleanUpdateFiles(rootPath);
			return;
		}
		logger->info("Unzipping was successful");

		// Now move the old files to outdated
		Navigator::GetInstance()->updateStatus = UpdateStatus::INSTALLING;
		logger->info("Moving old files to outdated");	// This might fail when another outdated executable is still running
		if (Battery::DirectoryExists(installPath)) {
			std::string target = rootPath + OUTDATED_DIRECTORY;
			for (int i = 0; i < 10; i++) {
				if (Battery::MoveDirectory(installPath, target)) {
					break;
				}
				
				target += "temp/";
			}
		}
		else {
			logger->info("Application can't be updated: Executable directory is inconsistent!"
				" Aborting to avoid breaking the installation...");
			Navigator::GetInstance()->timeSincePopup = Battery::GetRuntime();
			Navigator::GetInstance()->updateStatus = UpdateStatus::FAILED;
			CleanUpdateFiles(rootPath);
			return;
		}

		// Now move the new files to the install directory
		logger->info("Moving new files to rootdir");
		if (!Battery::MoveDirectory(rootPath + TEMP_DIRECTORY, rootPath)) {
			logger->info("MoveDirectory failed, reverting installation...");
			Battery::PrepareEmptyDirectory(installPath);
			Battery::MoveDirectory(rootPath + OUTDATED_DIRECTORY, installPath);
			logger->info("Tried my best to recover the old installation, aborting update");
			Navigator::GetInstance()->timeSincePopup = Battery::GetRuntime();
			Navigator::GetInstance()->updateStatus = UpdateStatus::FAILED;
			CleanUpdateFiles(rootPath);
			return;
		}
		if (installDir != "latest") {
			logger->info("Install directory was not called 'latest/', renaming to '{}/'", installDir);
			Battery::RemoveDirectory(installDir);
			Battery::RenameFile(rootPath + "latest/", rootPath + installDir);
		}
		

		// Delete update package
		logger->info("Cleaning update files");
		CleanUpdateFiles(rootPath);

		logger->info("Update was successful, stopping update thread");

		Navigator::GetInstance()->updateStatus = UpdateStatus::DONE;
		Navigator::GetInstance()->timeSincePopup = Battery::GetRuntime();

		// This .exe file is still running in the outdated folder:
		// Will be deleted when the application is opened the next time

	}
	catch (const Battery::Exception& e) {
		LOG_CORE_CRITICAL(std::string("The Updater thread threw Battery::Exception: ") + e.what());
		Battery::ShowErrorMessageBox(std::string("The Updater thread threw Battery::Exception: ") + e.what());
		Navigator::GetInstance()->updateStatus = UpdateStatus::NOTHING;
		return;
	}
	catch (...) {
		LOG_CORE_CRITICAL("The Updater thread threw unknown Exception: No further information");
		Battery::ShowErrorMessageBox("The Updater thread threw unknown Exception: No further information");
		Navigator::GetInstance()->updateStatus = UpdateStatus::NOTHING;
		return;
	}
}
